### PR TITLE
feat: redis cache should fallback to internal in case of error

### LINKED ISSUE #2788: Redis cache should fallback to internal on set and get operations
### Description

Currently, in the cacheService.ts the methods are written in a way that suggests that if an error occurs in Redis during set and getAsync operations we will fallback to the internal cache. However, this doesn't happen
For example
Screenshot 2024-08-05 at 14 10 08

  Here we will use the internalCache only if the shared one is not enabled

### Steps to reproduce

1. Check the cacheService.ts file code

### Additional context

_No response_

### Hedera network

other

### Version

0.52.1

### Operating system

None

### PR DESCRIPTION
**Pull request by natanasow**
**Description**:

Currently, in the cacheService.ts the methods are written in a way that suggests that if an error occurs in Redis during set and getAsync operations we will fallback to the internal cache. However, this doesn't happen.

**Related issue(s)**:

Fixes #2788

**Notes for reviewer**:

**Checklist**

- [ ] Documented (Code comments, README, etc.)
- [ ] Tested (unit, integration, etc.)

### CODE REVIEW

File: packages/relay/src/lib/services/cacheService/cacheService.ts [hunk 1]
```diff
@@ -216,10 +218,13 @@ export class CacheService {
...
+        // fallback to internal cache in case of Redis error
+        this.cacheMethodsCounter.labels(callingMethod, CacheService.cacheTypes.LRU, CacheService.methods.SET).inc(1);
+        await this.internalCache.set(key, value, callingMethod, ttl, requestIdPrefix);
       }
     } else {
       this.cacheMethodsCounter.labels(callingMethod, CacheService.cacheTypes.LRU, CacheService.methods.SET).inc(1);
-
       await this.internalCache.set(key, value, callingMethod, ttl, requestIdPrefix);
     }
   }
```
Comment (victor-yanev) [inline packages/relay/src/lib/services/cacheService/cacheService.ts:230]: **nit:** We can avoid duplicating the fallback logic inside the `catch` and `else` blocks by adding a `return` statement if the main logic finishes successfully:

```typescript
if (this.isSharedCacheEnabled) {
  try {
     this.cacheMethodsCounter.labels(...).inc(1);
     await this.sharedCache.set(...);
     return;
  } catch {
    this.logger.error('...');
  }
}
// fallback
this.cacheMethodsCounter.labels(...).inc(1);
await this.internalCache.set(...);
```

File: packages/relay/src/lib/services/cacheService/cacheService.ts [hunk 2]
```diff
@@ -289,6 +294,10 @@ export class CacheService {
         this.logger.error(
           `${requestIdPrefix} Error occurred while deleting cache from Redis. Error is: ${redisError.fullError}`,
         );
+
+        // fallback to internal cache in case of Redis error
+        this.cacheMethodsCounter.labels(callingMethod, CacheService.cacheTypes.LRU, CacheService.methods.DELETE).inc(1);
+        await this.internalCache.delete(key, callingMethod, requestIdPrefix);
```
Comment (victor-yanev) [inline packages/relay/src/lib/services/cacheService/cacheService.ts:300]: **nit:** Same here

File: packages/relay/src/lib/services/cacheService/cacheService.ts [hunk 3]
```diff
@@ -312,6 +321,9 @@ export class CacheService {
         this.logger.error(
           `${requestIdPrefix} Error occurred while clearing Redis cache. Error is: ${redisError.fullError}`,
         );
+
+        // fallback to internal cache in case of Redis error
+        await this.internalCache.clear();
```
Comment (victor-yanev) [inline packages/relay/src/lib/services/cacheService/cacheService.ts:326]: **nit:** Same here

File: packages/relay/tests/lib/services/cacheService/cacheService.spec.ts [hunk 4]
```diff
@@ -199,5 +199,45 @@ describe('CacheService Test Suite', async function () {
...
+
+      // @ts-ignore
+      cacheService.set.restore();
+
+      // @ts-ignore
+      mock.stub(cacheService.sharedCache, 'set').throwsException();
+      // @ts-ignore
+      mock.stub(cacheService.internalCache, 'set').returns(Promise<void>);
+
+      expect(await cacheService.set(key, callingMethod, undefined)).not.to.throw;
```
Comment (victor-yanev) [inline packages/relay/tests/lib/services/cacheService/cacheService.spec.ts:227]:
```suggestion
      await expect(cacheService.set(key, callingMethod)).to.eventually.not.be.rejected;
```

File: packages/relay/tests/lib/services/cacheService/cacheService.spec.ts [hunk 5]
```diff
@@ -199,5 +199,45 @@ describe('CacheService Test Suite', async function () {
...
+      const key = 'string';
+
+      // @ts-ignore
+      cacheService.delete.restore();
+      // @ts-ignore
+      mock.stub(cacheService.sharedCache, 'delete').throwsException();
+      // @ts-ignore
+      mock.stub(cacheService.internalCache, 'delete').returns(Promise<void>);
+
+      expect(await cacheService.delete(key, callingMethod, undefined)).not.to.throw;
```
Comment (victor-yanev) [inline packages/relay/tests/lib/services/cacheService/cacheService.spec.ts:240]:
```suggestion
      await expect(cacheService.delete(key, callingMethod)).to.eventually.not.be.rejected;
```

File: packages/relay/tests/lib/services/cacheService/cacheService.spec.ts [hunk 6]
```diff
@@ -199,5 +199,45 @@ describe('CacheService Test Suite', async function () {
...
+    it('should be able to getAsync from internal cache in case of Redis error', async function () {
+      const key = 'string';
+      const value = 'value';
+
+      // @ts-ignore
+      mock.stub(cacheService.sharedCache, 'get').throwsException();
+      // @ts-ignore
+      mock.stub(cacheService, 'getFromInternalCache').returns(value);
+
+      const cachedValue = await cacheService.getAsync(key, callingMethod, undefined);
```
Comment (victor-yanev) [inline packages/relay/tests/lib/services/cacheService/cacheService.spec.ts:212]:
```suggestion
      const cachedValue = await cacheService.getAsync(key, callingMethod);
```

File: packages/server/tests/acceptance/cacheService.spec.ts [hunk 7]
```diff
@@ -103,4 +103,59 @@ describe('@cache-service Acceptance Tests for shared cache', function () {
...
     expect(cachedData).to.deep.eq(DATA, 'cached data is read correctly by other service instance');
   });
+
+  describe('fallback to local cache in case of Redis error', async () => {
+    const dataLabel = `${DATA_LABEL_PREFIX}_redis_error`;
+
+    let currentRedisEnabledEnv;
+    let cacheService;
+
+    beforeEach(async () => {
```
Comment (victor-yanev) [inline packages/server/tests/acceptance/cacheService.spec.ts:113]: Let's not slow down the tests unnecessarily:

```suggestion
    before(async () => {
```

File: packages/server/tests/acceptance/cacheService.spec.ts [hunk 8]
```diff
@@ -103,4 +103,59 @@ describe('@cache-service Acceptance Tests for shared cache', function () {
...
+
+      process.env.REDIS_ENABLED = 'true';
+      cacheService = new CacheService(global.logger, registry);
+
+      // disconnect redis client to simulate Redis error
+      await cacheService.disconnectRedisClient();
+      await new Promise((r) => setTimeout(r, 1000));
+    });
+
+    afterEach(async () => {
```
Comment (victor-yanev) [inline packages/server/tests/acceptance/cacheService.spec.ts:124]:
```suggestion
    after(async () => {
```

**Review by victor-yanev — CHANGES_REQUESTED** [review]

**Review by konstantinabl — APPROVED** [review]

**Review by victor-yanev — APPROVED** [review]

**Pull request merged**